src/Nethermind/Nethermind.State/WorldState.cs

```diff
@@ -1,32 +1,460 @@
-// SPDX-FileCopyrightText: 2022 Demerzel Solutions Limited
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
 // SPDX-License-Identifier: LGPL-3.0-only
 
 using System;
+using System.Collections.Generic;
+using Nethermind.Core;
+using Nethermind.Core.Collections;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Extensions;
+using Nethermind.Core.Resettables;
+using Nethermind.Core.Specs;
+using Nethermind.Int256;
+using Nethermind.Logging;
+using Nethermind.Trie;
+using Nethermind.Trie.Pruning;
 
-namespace Nethermind.State
+namespace Nethermind.State;
+
+public class WorldState : PartialStorageProviderBase, IWorldState
```

Comment:
I don't like those class hierarchies a bit, now we have:

a) WorldState that covers Storage access
b) Inside it StateProvider for main tree
c) Inside it TransientStorage for transient storage
d) Remove `StorageProvider` and `PersistentStorageProvider` - probably to delete.

I would like to have it either:
a) World state covers everything 
b) World state delegates to Storage, TransientStorage, State internally

Now it is a mix, making it hard to understand.

Reply:
Earlier what I did first was something like this
- worldstate implemented IStateProvider and then the storage was delegated to StorageProvider. The thing I did not like about that was passing the WorldState in the StorageProvider.

The b) option also has the same issues (passing state provider in storage).

I will try to come up with something where WorldState implements everything.

Reply:
World state could internally delegate to State/Storage providers but hide them completely?

Reply:
I agree with what @LukaszRozmej proposes. Either delegate or derive. Half-delegate and derive does not look that nice. Still, interface-wise, that would be `IWorldState` that in case of `Verkle` can be implemented as one wants to.